### PR TITLE
Deployment returns outputs

### PR DESCRIPTION
**Pull request by tamirkamara**
# PR for issue #629 

## What is being addressed

Deployment via porter expose outputs we need to persist 

## How is this addressed

- Extract porter outputs and return in the deployment status message
- Change the way we prepare parameters for porter to allow for bundles without Terraform and also prioritize how the value sources
- Bundle meant to be deployed into workspaces (e.g. guacamole/vm) should be using the resource id as the Terraform store key
- Change guacamole/vm templates to make use special params (like workspace_id) automatically

### Additional
- Remove ANSI control chars from the shell outputs - will make it easier to read on app insights.

### CODE REVIEW

File: resource_processor/vmss_porter/runner.py [hunk 1]
```diff
@@ -82,40 +85,64 @@ async def filter_parameters_not_needed_by_porter(msg_body, env_vars):
...
 async def build_porter_command(msg_body, env_vars):
+    porter_parameter_keys = await get_porter_parameter_keys(msg_body, env_vars)
     porter_parameters = ""
 
-    porter_keys = await filter_parameters_not_needed_by_porter(msg_body, env_vars)
-    for parameter in porter_keys:
-        porter_parameters = porter_parameters + f" --param {parameter}=\"{msg_body['parameters'][parameter]}\""
+    for parameter_name in porter_parameter_keys:
+        # first try to find the param in the msg body parameters collection
+        parameter_value = msg_body['parameters'].get(parameter_name)
```
Comment (sachinkundu) [inline resource_processor/vmss_porter/runner.py:108]: parameter_value = msg_body['parameters'].get(parameter_name, env_vars.get(parameter_name, msg_body.get(parameter_name))) I think would do the same without all the checking for None.

File: resource_processor/vmss_porter/runner.py [hunk 2]
```diff
@@ -82,40 +85,64 @@ async def filter_parameters_not_needed_by_porter(msg_body, env_vars):
...
+        # if not found, try to get it from the msg body root (e.g. id of the resource)
+        if parameter_value is None:
+            parameter_value = msg_body.get(parameter_name)
 
-    porter_parameters = porter_parameters + f" --param tfstate_container_name={env_vars['tfstate_container_name']}"
-    porter_parameters = porter_parameters + f" --param tfstate_resource_group_name={env_vars['tfstate_resource_group_name']}"
-    porter_parameters = porter_parameters + f" --param tfstate_storage_account_name={env_vars['tfstate_storage_account_name']}"
-    porter_parameters = porter_parameters + f" --param arm_use_msi={env_vars['arm_use_msi']}"
+        # if still not found, might be a special case
+        if parameter_value is None:
```
Comment (sachinkundu) [inline resource_processor/vmss_porter/runner.py:119]: You know that the parameter_value is None so why not move this whole block at the start and break from the loop if parameter_name in ["mgmt_acr_name", "mgmt_resource_group_name"] as those are special. Infact could refactor method out to special_param_handling to make the intention clear.

**Review by sachinkundu — COMMENTED** [review]
Isn't the whole parameter thing UNION(param_set1, param_set2...).INTERSECT(porter_params)

File: resource_processor/vmss_porter/runner.py (hunk 2 above)
Comment (tamirkamara) [reply resource_processor/vmss_porter/runner.py]: I wanted to give an option for the api to supply those values in the msg params, which is why I left it at the end.

**Review by marrobi — APPROVED** [review]
Looks good. A few things that might affect existing bundles/we should be consistent across bundles.

Can you create an issue to sight on things like mgmt_acr_name, and any other can think of.

Thanks.

**Pull request merged**